File: internal/usagewatch/usagewatch.go
package usagewatch

import (
	"context"
	"fmt"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/phaedrus/curb/internal/config"
	"github.com/phaedrus/curb/internal/ledger"
	"github.com/phaedrus/curb/internal/platform"
	"github.com/phaedrus/curb/internal/usage"
	"github.com/phaedrus/curb/internal/watchdog"
)

type EventReader func(time.Time) ([]usage.Event, []usage.SourceReport, error)
type Capture func(context.Context) (*platform.Snapshot, error)
type Notify func(string, string) error
type Terminate func(context.Context, platform.TerminationTarget, time.Duration) platform.TerminationResult

type Service struct {
	cfg       *config.Config
	ledger    *ledger.Ledger
	reader    EventReader
	capture   Capture
	notify    Notify
	terminate Terminate
	now       func() time.Time
	onEvent   func(ledger.Event)
	warned    map[string]bool
	grace     map[string]time.Time
	targets   map[string]platform.Process
}

type Session struct {
	Key            string
	Provider       string
	SessionID      string
	CWD            string
	Last           time.Time
	LastUsage      time.Time
	Events         int
	Input          int64
	CachedInput    int64
	CacheCreation  int64
	Output         int64
	Reasoning      int64
	Total          int64
	LastTurnTokens int64
	Models         []string
}

type PolicyResult struct {
	State       string
	Active      bool
	WindowStart time.Time
	Explanation string
}

type SessionDecision struct {
	State       string
	UsageState  string
	Actionable  bool
	Policy      PolicyResult
	Explanation string
}

type SessionClassification struct {
	State          string
	ProcessState   string
	UsageState     string
	ActionState    string
	Actionable     bool
	CanAcknowledge bool
	RiskRank       int
	Explanation    string
}

type Correlation struct {
	Matched    bool
	Agent      config.Agent
	Process    platform.Process
	Score      int
	Reason     string
	Confidence int
	Evidence   []string
}

func New(cfg *config.Config, l *ledger.Ledger) *Service {
	return &Service{
		cfg:       cfg,
		ledger:    l,
		reader:    usage.EventsSince,
		capture:   platform.Capture,
		notify:    platform.Notify,
		terminate: platform.TerminateTree,
		now:       time.Now,
		warned:    map[string]bool{},
		grace:     map[string]time.Time{},
		targets:   map[string]platform.Process{},
	}
}

func (s *Service) OnEvent(fn func(ledger.Event)) {
	s.onEvent = fn
}

func (s *Service) SetCapture(capture Capture) {
	if capture != nil {
		s.capture = capture
	}
}

func (s *Service) SetNotify(notify Notify) {
	if notify != nil {
		s.notify = notify
	}
}

func (s *Service) SetTerminate(terminate Terminate) {
	if terminate != nil {
		s.terminate = terminate
	}
}

func (s *Service) SetReader(reader EventReader) {
	if reader != nil {
		s.reader = reader
	}
}

func (s *Service) Reconfigure(cfg *config.Config, l *ledger.Ledger) {
	s.cfg = cfg
	s.ledger = l
}

func (s *Service) Run(ctx context.Context) error {
	if !s.cfg.Usage.IsEnabled() {
		return nil
	}
	ticker := time.NewTicker(s.cfg.Usage.ScanInterval.Duration)
	defer ticker.Stop()
	if err := s.Scan(ctx); err != nil {
		return err
	}
	for {
		select {
		case <-ctx.Done():
			return nil
		case <-ticker.C:
			if err := s.Scan(ctx); err != nil {
				_ = s.append(ledger.Event{Type: "usage_scan_failed", Mode: string(s.cfg.Mode), Message: err.Error()})
			}
		}
	}
}

func (s *Service) Scan(ctx context.Context) error {
	now := s.now()
	events, _, err := s.reader(now.Add(-s.cfg.Usage.Lookback.Duration))
	if err != nil {
		return err
	}
	sessions := BuildSessions(events)
	if len(sessions) == 0 {
		return nil
	}
	snap, err := s.capture(ctx)
	if err != nil {
		return err
	}
	matches := processMatches(s.cfg, s.ledger, snap)
	for _, session := range sessions {
		policy := EvaluateSessionPolicy(session, s.cfg.Usage, now)
		if policy.State != "warn" && policy.State != "stop" {
			continue
		}
		_, ok, err := ActiveSessionAck(s.cfg.Service.StateDir, session.Key, now)
		if err != nil {
			_ = s.append(ledger.Event{Type: "usage_ack_failed", Mode: string(s.cfg.Mode), Message: err.Error(), Data: map[string]any{"session_key": session.Key}})
		}
		if ok {
			s.suppressUntilAckExpires(session.Key)
			continue
		}
		correlation := Correlate(session, matches)
		if err := s.evaluate(ctx, snap, session, correlation, now); err != nil {
			return err
		}
	}
	return nil
}

func BuildSessions(events []usage.Event) []Session {
	byKey := map[string]*Session{}
	models := map[string]map[string]bool{}
	for _, event := range events {
		key := SessionKey(event.Provider, event.SessionID, event.SourcePath)
		session := byKey[key]
		if session == nil {
			session = &Session{Key: key, Provider: event.Provider, SessionID: event.SessionID, CWD: event.CWD}
			byKey[key] = session
			models[key] = map[string]bool{}
		}
		session.Events++
		session.Input += event.Input
		session.CachedInput += event.CachedInput
		session.CacheCreation += event.CacheCreation
		session.Output += event.Output
		session.Reasoning += event.Reasoning
		session.Total += event.Total
		if !event.Timestamp.Before(session.Last) {
			session.Last = event.Timestamp
		}
		if event.Total > 0 && !event.Timestamp.Before(session.LastUsage) {
			session.LastUsage = event.Timestamp
			session.LastTurnTokens = event.Total
		}
		if session.CWD == "" {
			session.CWD = event.CWD
		}
		if event.Model != "" {
			models[key][event.Model] = true
		}
	}
	out := make([]Session, 0, len(byKey))
	for key, session := range byKey {
		for model := range models[key] {
			session.Models = append(session.Models, model)
		}
		sort.Strings(session.Models)
		out = append(out, *session)
	}
	sort.Slice(out, func(i, j int) bool { return out[i].Last.After(out[j].Last) })
	return out
}

func SessionKey(provider, sessionID, sourcePath string) string {
	if sessionID != "" {
		return provider + ":" + sessionID
	}
	return provider + ":" + sourcePath
}

func (s Session) RecentUsage(windowStart time.Time) bool {
	return !s.LastUsage.IsZero() && !s.LastUsage.Before(windowStart)
}

func EvaluateSessionPolicy(session Session, usageCfg config.UsageConfig, now time.Time) PolicyResult {
	if now.IsZero() {
		now = time.Now()
	}
	windowStart := now.Add(-usageCfg.Window.Duration)
	active := session.RecentUsage(windowStart)
	result := PolicyResult{
		State:       "idle",
		Active:      active,
		WindowStart: windowStart,
		Explanation: "no usage in current window",
	}
	if active {
		result.State = "active"
		result.Explanation = "latest turn within limits"
	}
	switch {
	case active && session.LastTurnTokens >= usageCfg.KillTurnTokens:
		result.State = "stop"
		result.Explanation = "latest turn over stop threshold"
	case active && session.LastTurnTokens >= usageCfg.WarnTurnTokens:
		result.State = "warn"
		result.Explanation = "latest turn over warning threshold"
	case !active && session.LastTurnTokens >= usageCfg.KillTurnTokens:
		result.State = "idle-high"
		result.Explanation = "large historical turn, not currently spending"
	}
	return result
}

func Correlate(session Session, matches []watchdog.Match) Correlation {
	var best Correlation
	sessionCWD := cleanPath(session.CWD)
	if sessionCWD == "" {
		return best
	}
	for _, match := range matches {
		if !sameProvider(session.Provider, match.Agent.Family) {
			continue
		}
		processCWD := cleanPath(match.Process.CWD)
		if processCWD == "" {
			continue
		}
		score := 25
		reason := "provider"
		switch {
		case processCWD == sessionCWD:
			score += 100
			reason = "provider+cwd"
		case pathContains(processCWD, sessionCWD) || pathContains(sessionCWD, processCWD):
			score += 50
			reason = "provider+cwd-prefix"
		default:
			continue
		}
		if score > best.Score {
			best = Correlation{
				Matched:    true,
				Agent:      match.Agent,
				Process:    match.Process,
				Score:      score,
				Reason:     reason,
				Confidence: match.Confidence,
				Evidence:   append([]string(nil), match.Evidence...),
			}
		}
	}
	return best
}

func EvaluateSessionDecision(session Session, cfg *config.Config, correlation Correlation, now time.Time) SessionDecision {
	policy := EvaluateSessionPolicy(session, cfg.Usage, now)
	decision := SessionDecision{
		State:       policy.State,
		UsageState:  policy.State,
		Policy:      policy,
		Explanation: policy.Explanation,
	}
	if policy.State == "idle" {
		decision.UsageState = ""
		return decision
	}
	if policy.State != "warn" && policy.State != "stop" {
		return decision
	}
	switch {
	case !correlation.Matched:
		decision.State = "uncorrelated"
		decision.Explanation = "usage crossed threshold, but no live process matched; Curb will not stop anything"
	case !correlation.Agent.TerminationAllowed():
		decision.State = "watch-only"
		decision.Explanation = "usage crossed threshold, but matched agent is watch-only; Curb will not stop desktop apps"
	case cfg.Mode == config.ModeEnforcement && policy.State == "stop":
		decision.Actionable = true
	}
	return decision
}

func ClassifySession(decision SessionDecision, correlation Correlation, mode config.Mode, acknowledgedUntil *time.Time, ackExtension time.Duration) SessionClassification {
	classification := SessionClassification{
		State:       decision.State,
		UsageState:  sessionUsageState(decision),
		Actionable:  decision.Actionable,
		Explanation: decision.Explanation,
	}
	if acknowledgedUntil != nil && (classification.UsageState == "warn" || classification.UsageState == "stop") {
		classification.State = "acknowledged"
		classification.Actionable = false
		classification.Explanation = "usage crossed threshold, but this session is acknowledged until " + acknowledgedUntil.Format(time.RFC3339)
	}
	classification.ProcessState = sessionProcessState(classification, correlation)
	classification.ActionState = sessionActionState(classification, mode, correlation, acknowledgedUntil != nil)
	classification.CanAcknowledge = acknowledgedUntil == nil && ackExtension > 0 && (classification.UsageState == "warn" || classification.UsageState == "stop")
	classification.RiskRank = sessionRiskRank(classification)
	return classification
}

func sessionProcessState(classification SessionClassification, correlation Correlation) string {
	switch {
	case classification.State == "watch-only":
		return "watch-only"
	case correlation.Matched:
		return "running"
	case classification.State == "uncorrelated" || classification.UsageState == "warn" || classification.UsageState == "stop":
		return "unknown"
	default:
		return "no-process"
	}
}

func sessionUsageState(decision SessionDecision) string {
	switch {
	case decision.UsageState == "warn" || decision.UsageState == "stop":
		return decision.UsageState
	case decision.State == "warn" || decision.State == "stop":
		return decision.State
	case decision.State == "active":
		return "spending"
	case decision.State == "idle-high":
		return "quiet-high"
	default:
		return "quiet"
	}
}

func sessionActionState(classification SessionClassification, mode config.Mode, correlation Correlation, acknowledged bool) string {
	if acknowledged {
		return "acknowledged"
	}
	if classification.UsageState == "stop" || classification.State == "stop" {
		if classification.Actionable {
			return "stop-pending"
		}
		if mode != config.ModeEnforcement && correlation.Matched && classification.ProcessState == "running" {
			return "would-stop"
		}
		return "blocked"
	}
	if classification.UsageState == "warn" || classification.State == "warn" {
		return "acknowledge"
	}
	return "none"
}

func sessionRiskRank(classification SessionClassification) int {
	switch {
	case classification.Actionable:
		return 0
	case classification.State == "acknowledged":
		return 2
	case classification.State == "idle-high":
		return 3
	case classification.UsageState == "stop" || classification.UsageState == "warn" || classification.UsageState == "spending" || classification.State == "active":
		return 1
	case classification.ProcessState == "running":
		return 4
	default:
		return 5
	}
}

func (s *Service) suppressUntilAckExpires(sessionKey string) {
	s.warned[sessionKey] = false
	s.warned["would:"+sessionKey] = false
	s.warned["uncorrelated:"+sessionKey] = false
	delete(s.grace, sessionKey)
	delete(s.targets, sessionKey)
}

func (s *Service) evaluate(ctx context.Context, snap *platform.Snapshot, session Session, correlation Correlation, now time.Time) error {
	key := session.Key
	policy := EvaluateSessionPolicy(session, s.cfg.Usage, now)
	overKill := policy.State == "stop"
	msg := fmt.Sprintf("%s session %s latest turn used %s tokens (total %s in %d calls)",
		session.Provider, shortID(session.SessionID), formatTokens(session.LastTurnTokens), formatTokens(session.Total), session.Events)

	if !s.warned[key] {
		s.warned[key] = true
		if s.cfg.Alerts.LocalNotifications {
			if err := s.notify("Curb usage warning", msg); err != nil {
				_ = s.append(ledger.Event{Type: "notification_failed", Mode: string(s.cfg.Mode), Message: err.Error()})
			}
		}
		if err := s.append(ledger.Event{
			Type:    "usage_warning",
			AgentID: correlation.Agent.ID,
			Mode:    string(s.cfg.Mode),
			Message: msg,
			Data:    eventData(session, correlation),
		}); err != nil {
			return err
		}
	}
	if !overKill {
		return nil
	}
	if !correlation.Matched {
		if !s.warned["uncorrelated:"+key] {
			s.warned["uncorrelated:"+key] = true
			return s.append(ledger.Event{
				Type:    "usage_kill_blocked",
				Mode:    string(s.cfg.Mode),
				Message: "usage threshold exceeded but no live process correlation was found",
				Data:    eventData(session, correlation),
			})
		}
		return nil
	}
	if !correlation.Agent.TerminationAllowed() {
		return s.append(ledger.Event{
			Type:    "usage_kill_blocked",
			AgentID: correlation.Agent.ID,
			Mode:    string(s.cfg.Mode),
			Message: "usage threshold exceeded but matched agent is watch-only",
			Data:    eventData(session, correlation),
		})
	}
	if s.cfg.Mode != config.ModeEnforcement {
		if !s.warned["would:"+key] {
			s.warned["would:"+key] = true
			if s.cfg.Alerts.LocalNotifications {
				_ = s.notify("Curb would stop agent", msg)
			}
			return s.append(ledger.Event{
				Type:    "usage_would_terminate",
				AgentID: correlation.Agent.ID,
				Mode:    string(s.cfg.Mode),
				Message: msg,
				Data:    eventData(session, correlation),
			})
		}
		return nil
	}
	started := s.grace[key]
	if started.IsZero() {
		s.grace[key] = now
		s.targets[key] = correlation.Process
		if s.cfg.Alerts.LocalNotifications {
			_ = s.notify("Curb usage grace period", msg)
		}
		return s.append(ledger.Event{
			Type:    "usage_grace_started",
			AgentID: correlation.Agent.ID,
			Mode:    string(s.cfg.Mode),
			Message: msg,
			Data:    eventData(session, correlation),
		})
	}
	if now.Sub(started) < s.cfg.Usage.GracePeriod.Duration {
		return nil
	}

	target := correlation.Process
	if stored, ok := s.targets[key]; ok {
		target = stored
	}
	terminationCorrelation := correlation
	terminationCorrelation.Process = target
	terminationTarget, ok := snap.TerminationTarget(target)
	if !ok {
		return s.append(ledger.Event{
			Type:    "usage_termination_failed",
			AgentID: correlation.Agent.ID,
			Mode:    string(s.cfg.Mode),
			Message: "safety guard rejected termination",
			Data:    eventData(session, terminationCorrelation),
		})
	}
	if err := s.append(ledger.Event{
		Type:    "usage_termination_started",
		AgentID: correlation.Agent.ID,
		Mode:    string(s.cfg.Mode),
		Message: msg,
		Data:    eventData(session, terminationCorrelation),
	}); err != nil {
		return err
	}
	result := s.terminate(ctx, terminationTarget, s.cfg.Usage.GracePeriod.Duration)
	data := eventData(session, terminationCorrelation)
	data["result"] = result
	return s.append(ledger.Event{
		Type:    "usage_termination_completed",
		AgentID: correlation.Agent.ID,
		Mode:    string(s.cfg.Mode),
		Message: msg,
		Data:    data,
	})
}

func (s *Service) append(event ledger.Event) error {
	if s.ledger != nil {
		if err := s.ledger.Append(event); err != nil {
			return err
		}
	}
	if s.onEvent != nil {
		s.onEvent(event)
	}
	return nil
}

func processMatches(cfg *config.Config, l *ledger.Ledger, snap *platform.Snapshot) []watchdog.Match {
	service := watchdog.New(cfg, l)
	return service.Match(snap)
}

func eventData(session Session, correlation Correlation) map[string]any {
	data := map[string]any{
		"provider":     session.Provider,
		"session_id":   session.SessionID,
		"cwd":          session.CWD,
		"calls":        session.Events,
		"total_tokens": session.Total,
		"turn_tokens":  session.LastTurnTokens,
		"last":         session.Last.UTC(),
	}
	if !session.LastUsage.IsZero() {
		data["last_usage"] = session.LastUsage.UTC()
	}
	if len(session.Models) > 0 {
		data["models"] = session.Models
	}
	if correlation.Matched {
		data["pid"] = correlation.Process.PID
		data["agent_id"] = correlation.Agent.ID
		data["correlation"] = correlation.Reason
		data["correlation_score"] = correlation.Score
	}
	return data
}

func sameProvider(provider, family string) bool {
	return strings.EqualFold(provider, family)
}

func cleanPath(path string) string {
	if path == "" {
		return ""
	}
	if abs, err := filepath.Abs(path); err == nil {
		path = abs
	}
	return filepath.Clean(path)
}

func pathContains(parent, child string) bool {
	if parent == "" || child == "" || parent == child {
		return false
	}
	if !strings.HasPrefix(child, parent) {
		return false
	}
	if strings.HasSuffix(parent, string(filepath.Separator)) {
		return true
	}
	return len(child) > len(parent) && child[len(parent)] == filepath.Separator
}

func shortID(id string) string {
	if len(id) <= 12 {
		return id
	}
	return id[:8] + "..." + id[len(id)-4:]
}

func formatTokens(n int64) string {
	if n >= 1_000_000 {
		return fmt.Sprintf("%.1fM", float64(n)/1_000_000)
	}
	if n >= 10_000 {
		return fmt.Sprintf("%dk", n/1_000)
	}
	return fmt.Sprintf("%d", n)
}

File: curb-core/src/usagewatch.rs
use std::collections::{BTreeSet, HashMap, HashSet};
use std::path::PathBuf;

use chrono::{DateTime, Utc};
use serde_json::Value;
use thiserror::Error;

use crate::config::{Config, Mode};
use crate::ledger::{self, Ledger, LedgerEvent};

mod events;
#[cfg(test)]
mod tests;

#[derive(Debug, Error)]
pub enum UsageWatchError {
    #[error(transparent)]
    Ledger(#[from] ledger::LedgerError),
}

/// A correlated session the policy evaluates, free of OS process facts. The
/// caller (runtime/e2e) builds these from raw usage events; the policy core
/// never sees `usage::Event`, `platform::Snapshot`, or `platform::Process`.
#[derive(Clone, Debug)]
pub struct PolicySession {
    pub key: String,
    pub id: String,
    pub provider: String,
    pub cwd: Option<PathBuf>,
    pub models: BTreeSet<String>,
    pub last: Option<DateTime<Utc>>,
    pub last_usage: Option<DateTime<Utc>>,
    pub calls: usize,
    pub latest_turn_tokens: i64,
    pub latest_spent_tokens: i64,
    pub window_spent_tokens: i64,
    pub total_tokens: i64,
    /// The correlation the caller resolved for this session.
    pub target: AgentTarget,
    /// Whether the operator has an active acknowledgement suppressing this
    /// session. The caller reads the ack store; the policy stays I/O-free.
    pub acknowledged: bool,
}

impl PolicySession {
    fn recent_usage(&self, window_start: DateTime<Utc>) -> bool {
        self.last_usage
            .is_some_and(|last_usage| last_usage >= window_start)
    }
}

/// An opaque token the enforcer can revalidate and stop. The policy stores it
/// across scans (grace lifecycle) without ever inspecting its contents; only
/// the enforcer that produced it knows how to resolve it back to a live target.
/// The OS seal (pid + start + owner + executable) lives inside the concrete
/// token an enforcer downcasts to — never in the policy core.
pub trait StopToken: std::any::Any + std::fmt::Debug + Send {
    fn clone_token(&self) -> Box<dyn StopToken>;
    fn as_any(&self) -> &dyn std::any::Any;
}

impl Clone for Box<dyn StopToken> {
    fn clone(&self) -> Self {
        self.clone_token()
    }
}

/// The pre-correlated, environment-agnostic view of a session's worker. Carries
/// only what the policy needs to decide — never an OS `Process`.
#[derive(Clone, Debug, Default)]
pub struct AgentTarget {
    pub matched: bool,
    pub agent_id: Option<String>,
    /// `true` when an agent matched and Curb may terminate it under the active
    /// escalation setting. Resolved by the caller from `Agent::can_terminate`.
    pub can_terminate: bool,
    /// `true` when the matched agent is a supervised desktop worker. Drives the
    /// escalation decision and the watch-only messaging.
    pub supervised: bool,
    /// The live worker pid, for ledger projection. `None` when uncorrelated.
    pub pid: Option<i64>,
    pub score: i64,
    pub reason: String,
    /// The token the enforcer uses to revalidate and stop the worker, captured
    /// at correlation time. `None` when uncorrelated.
    pub stop_token: Option<Box<dyn StopToken>>,
}

/// The outcome of an [`Enforcer::stop`] attempt, projected into the ledger.
pub enum StopResolution {
    /// The safety guard resolved a live target and the stop ran. Carries the
    /// already-serialized termination result for the completed ledger event.
    Stopped(Value),
    /// The safety guard rejected the stop (e.g. pid reuse). Nothing died.
    Rejected,
}

/// The side-effecting actions the policy delegates. The local implementation
/// owns the OS specifics (the sealed termination target, supervisor escalation,
/// the kill primitive); a remote implementation governs its own world.
pub trait Enforcer {
    /// Deliver an operator notification. Failures are the enforcer's concern.
    fn notify(&self, title: &str, message: &str);
    /// Revalidate and stop the worker behind `token`. `escalate` requests the
    /// supervisor's tree instead of the leaf for supervised desktop workers.
    fn stop(&self, token: &dyn StopToken, escalate: bool) -> StopResolution;
}

#[derive(Clone, Debug, Default)]
pub struct UsageWatch {
    warned: HashSet<String>,
    grace: HashMap<String, DateTime<Utc>>,
    targets: HashMap<String, Box<dyn StopToken>>,
    /// Sessions whose worker Curb has terminated, keyed to the kill time. The
    /// read model drops these rows so a killed agent leaves the dashboard at
    /// once instead of lingering on log recency, and the scan stops re-warning
    /// or re-killing them. Cleared when the session resumes (new activity after
    /// the kill) or ages out of the window.
    terminated: HashMap<String, DateTime<Utc>>,
}

/// Machine-readable summary of one policy scan.
///
/// The ledger remains the durable audit trail. This report is the compact
/// per-tick shape runtime observability can log without scraping the ledger or
/// inferring grace/stop outcomes from UI prose.
#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct PolicyScanReport {
    pub observed_sessions: usize,
    pub warnings: usize,
    pub would_stop: usize,
    pub stop_blocked: usize,
    pub grace_started: usize,
    pub grace_pending: usize,
    pub stop_attempted: usize,
    pub stop_completed: usize,
    pub stop_rejected: usize,
    pub resumed_sessions: usize,
    pub terminated_sessions: usize,
}

impl UsageWatch {
    /// Evaluate the correlated `sessions` against config and drive the enforcer.
    /// Pure policy: the caller has already built sessions, resolved correlation
    /// and acks, and supplied the `enforcer`; this method owns only thresholds,
    /// the grace/terminated state machine, and the ledger projection.
    pub fn scan<E: Enforcer>(
        &mut self,
        cfg: &Config,
        sessions: &[PolicySession],
        enforcer: &E,
        window_start: DateTime<Utc>,
        now: DateTime<Utc>,
    ) -> Result<PolicyScanReport, UsageWatchError> {
        let mut report = PolicyScanReport {
            observed_sessions: sessions.len(),
            ..PolicyScanReport::default()
        };
        if !cfg.usage.enabled() {
            return Ok(report);
        }
        if sessions.is_empty() {
            self.clear();
            return Ok(report);
        }
        // Forget kills that have aged out of the window — beyond it the session
        // is a finished run the read model drops on recency anyway.
        self.terminated
            .retain(|_, killed_at| *killed_at >= window_start);
        let mut active_keys = BTreeSet::new();
        for session in sessions {
            if !session.recent_usage(window_start)
                || session.latest_spent_tokens < cfg.usage.warn_turn_tokens
            {
                self.suppress(&session.key);
                continue;
            }
            // A killed session that has logged no new activity is still dead —
            // skip it so Curb does not re-warn, re-kill, or spam "stop blocked".
            // Fresh activity after the kill means it came back: clear and re-arm.
            if let Some(killed_at) = self.terminated.get(&session.key).copied() {
                if session.last_usage.is_some_and(|last| last > killed_at) {
                    self.terminated.remove(&session.key);
                    report.resumed_sessions += 1;
                } else {
                    continue;
                }
            }
            active_keys.insert(session.key.clone());
            if session.acknowledged {
                self.suppress(&session.key);
                continue;
            }
            self.evaluate(cfg, session, enforcer, now, &mut report)?;
        }
        self.retain_active(&active_keys);
        report.terminated_sessions = self.terminated.len();
        Ok(report)
    }

    fn evaluate<E: Enforcer>(
        &mut self,
        cfg: &Config,
        session: &PolicySession,
        enforcer: &E,
        now: DateTime<Utc>,
        report: &mut PolicyScanReport,
    ) -> Result<(), UsageWatchError> {
        let key = session.key.as_str();
        let target = &session.target;
        let over_stop = session.latest_spent_tokens >= cfg.usage.kill_turn_tokens;
        let message = events::usage_message(session);

        if self.warned.insert(key.to_string()) {
            notify_user(cfg, enforcer, "Curb usage warning", &message);
            append_event(cfg, LedgerEvent::UsageWarning, session, &message, None)?;
            report.warnings += 1;
        }
        if !over_stop {
            self.grace.remove(key);
            self.targets.remove(key);
            return Ok(());
        }
        if !target.matched {
            let blocked_key = format!("uncorrelated:{key}");
            if self.warned.insert(blocked_key) {
                notify_user(
                    cfg,
                    enforcer,
                    "Curb stop blocked",
                    "Usage threshold exceeded, but Curb could not correlate this session to a live worker.",
                );
                append_event(
                    cfg,
                    LedgerEvent::UsageKillBlocked,
                    session,
                    "usage threshold exceeded but no live process correlation was found",
                    None,
                )?;
                report.stop_blocked += 1;
            }
            return Ok(());
        }
        if target.agent_id.is_none() {
            return Ok(());
        }
        if !target.can_terminate {
            let blocked_key = format!("watch-only:{key}");
            if self.warned.insert(blocked_key) {
                let (title, detail) = if target.supervised {
                    (
                        "Curb can't stop this agent",
                        "Over the kill line, but a desktop app supervises this task and would respawn it. Enable escalate_supervised to stop it.",
                    )
                } else {
                    (
                        "Curb stop blocked",
                        "Usage threshold exceeded, but the matched process is watch-only.",
                    )
                };
                notify_user(cfg, enforcer, title, detail);
                append_event(cfg, LedgerEvent::UsageKillBlocked, session, detail, None)?;
                report.stop_blocked += 1;
            }
            return Ok(());
        }
        if cfg.mode != Mode::Enforcement {
            let would_key = format!("would:{key}");
            if self.warned.insert(would_key) {
                notify_user(cfg, enforcer, "Curb would stop agent", &message);
                append_event(
                    cfg,
                    LedgerEvent::UsageWouldTerminate,
                    session,
                    &message,
                    None,
                )?;
                report.would_stop += 1;
            }
            return Ok(());
        }

        let Some(stop_token) = &target.stop_token else {
            return Ok(());
        };
        if !self.grace.contains_key(key) {
            self.grace.insert(key.to_string(), now);
            self.targets.insert(key.to_string(), stop_token.clone());
            notify_user(cfg, enforcer, "Curb usage grace period", &message);
            append_event(cfg, LedgerEvent::UsageGraceStarted, session, &message, None)?;
            report.grace_started += 1;
            return Ok(());
        }
        let started = self.grace[key];
        if now.signed_duration_since(started)
            < chrono::Duration::from_std(cfg.usage.grace_period.as_std()).unwrap()
        {
            report.grace_pending += 1;
            return Ok(());
        }

        // Stop the grace-time target, falling back to the current correlation if
        // none was stored. Supervised desktop workers respawn when their leaf is
        // killed; with the escalate opt-in we target the supervisor's tree.
        let stored = self.targets.get(key).cloned();
        let stop_target: &dyn StopToken = match &stored {
            Some(token) => token.as_ref(),
            None => stop_token.as_ref(),
        };
        let escalate = target.supervised && cfg.usage.escalate_supervised;
        report.stop_attempted += 1;
        // The identity seal and the kill are bundled inside the env-agnostic
        // `Enforcer::stop`, so the termination_started/completed ledger pair is
        // written once the stop resolves. The event *sequence* (started ->
        // completed, or a lone failed) matches the old inline platform path; the
        // policy just no longer holds OS concepts to resolve the target itself.
        match enforcer.stop(stop_target, escalate) {
            StopResolution::Rejected => {
                notify_user(
                    cfg,
                    enforcer,
                    "Curb stop failed",
                    "Safety guard rejected termination for a stop-pending session.",
                );
                append_event(
                    cfg,
                    LedgerEvent::UsageTerminationFailed,
                    session,
                    "safety guard rejected termination",
                    None,
                )?;
                report.stop_rejected += 1;
            }
            StopResolution::Stopped(result) => {
                append_event(
                    cfg,
                    LedgerEvent::UsageTerminationStarted,
                    session,
                    &message,
                    None,
                )?;
                notify_user(cfg, enforcer, "Curb stopped agent", &message);
                append_event(
                    cfg,
                    LedgerEvent::UsageTerminationCompleted,
                    session,
                    &message,
                    Some(result),
                )?;
                report.stop_completed += 1;
                // Mark the session killed so the read model drops its row
                // immediately and the next scan stops re-warning or re-killing.
                self.terminated.insert(key.to_string(), now);
                self.grace.remove(key);
                self.targets.remove(key);
            }
        }
        Ok(())
    }

    /// Sessions Curb has terminated and that have not resumed — the read model
    /// drops their rows so a killed agent leaves the dashboard at once.
    pub fn terminated_keys(&self) -> BTreeSet<String> {
        self.terminated.keys().cloned().collect()
    }

    fn suppress(&mut self, key: &str) {
        self.warned.remove(key);
        self.warned.remove(&format!("would:{key}"));
        self.warned.remove(&format!("uncorrelated:{key}"));
        self.warned.remove(&format!("watch-only:{key}"));
        self.grace.remove(key);
        self.targets.remove(key);
        self.terminated.remove(key);
    }

    fn retain_active(&mut self, active_keys: &BTreeSet<String>) {
        self.warned.retain(|key| {
            active_keys.contains(
                key.strip_prefix("would:")
                    .or_else(|| key.strip_prefix("uncorrelated:"))
                    .or_else(|| key.strip_prefix("watch-only:"))
                    .unwrap_or(key),
            )
        });
        self.grace.retain(|key, _| active_keys.contains(key));
        self.targets.retain(|key, _| active_keys.contains(key));
    }

    fn clear(&mut self) {
        self.warned.clear();
        self.grace.clear();
        self.targets.clear();
        self.terminated.clear();
    }
}

fn notify_user<E: Enforcer>(cfg: &Config, enforcer: &E, title: &str, message: &str) {
    if !cfg.alerts.local_notifications {
        return;
    }
    enforcer.notify(title, message);
}

fn append_event(
    cfg: &Config,
    event_type: LedgerEvent,
    session: &PolicySession,
    message: &str,
    result: Option<Value>,
) -> Result<(), UsageWatchError> {
    let mut event = ledger::Event::new(event_type.as_str())
        .with_message(message.to_string())
        .with_data(events::event_data(session, result));
    event.agent_id = session.target.agent_id.clone();
    event.mode = Some(cfg.mode.to_string());
    Ledger::open(&cfg.ledger.path)?.append(event)?;
    Ok(())
}
